### ISSUE TITLE
UoW loses most of the collection changes 

### ISSUE THREAD

**Issue by derevnjuk** [mikro-orm/mikro-orm#3601]
**Describe the bug**

UoW loses most of the collection updates while computing changes on flush. It looks like regression due to 0fb17bbb

**Steps to reproduce**
1. Declare two simple entities as follows:
```ts
@Entity()
export class Group {
  @PrimaryKey({columnType: 'uuid', defaultRaw: `uuid_generate_v4()`})
  public id: string = v4();

  constructor(params: EntityData<Group>) {
    Object.assign(this, params);
  }
}

@Entity()
export class Member {
  @PrimaryKey({columnType: 'uuid', defaultRaw: `uuid_generate_v4()`})
  public id: string = v4();

  constructor(params: EntityData<Member>) {
    Object.assign(this, params);
  }
}
```
2. Declare an entity, representing a many-to-many relation between previously created entities (i.e. `Group` and `Member`)

```ts
@Entity()
export class GroupMember {
  @ManyToOne({
    entity: () => Member,
    primary: true,
    wrappedReference: true,
  })
  public member: IdentifiedReference<Member>;

  @ManyToOne({
    entity: () => Group,
    primary: true,
    wrappedReference: true,
  })
  public group: IdentifiedReference<Group>;

  public [PrimaryKeyType]!: [string, string];
  public [PrimaryKeyProp]!: 'member' | 'group';

  constructor(params: EntityData<GroupMember>) {
    Object.assign(this, params);
  }
}
```
3. Update them to establish the inverse side of many-to-one relation
4. Insert multiple records to the pivot table, using previously created instances of member and group, like this:
```ts
const group1 = new Group({});
const group2 = new Group({});
const member = new Member({});

await orm.em.persistAndFlush([group1, group2, member]);

orm.em.assign(member, {
  groups: [
    {
      group: group2.id,
      member: member.id,
    },
    {
      group: group1.id,
      member: member.id
    }
  ]
});

await orm.em.persistAndFlush(member);
```
5. Remove one of the added entities as follows:
```ts
orm.em.assign(member, {
  groups: [
    {
      group: group1.id,
      member: member.id,
    },
  ],
});

await orm.em.persistAndFlush(member);
```

**Expected behavior**: UoW tacks all the changes preserving the order of generated inserts statements in accordance with relations. `EntityManager` generates the following queries:

```sql
insert into "group_member" ("group_id", "member_id") values ($1, $2)
insert into "group_member" ("group_id", "member_id") values ($1, $2)
delete from "group_member" where ("member_id", "group_id") in (($1, $2))
```

**Actual behavior**: UoW loses most of the collection updates while computing changes on flush, mistakenly marking new entities as managed. `EntityManager` generates the following query:
```sql
delete from "group_member" where ("member_id", "group_id") in (($1, $2))
```

**Additional context**
There is repro-repo here: https://github.com/derevnjuk/mikro-orm-gh3599 (update mikro-orm up to 5.4.3-dev.21)

**Versions**

| Dependency | Version |
| - | - |
| node | >= 14 |
| typescript | >=4 |
| mikro-orm | >5.4.2 |

**Comment by B4nan** [comment]
This seems to be a problem with the `assign` helper, if you replace the assign calls with `member.groups.set()` it works as expected.

FYI you should be careful with `Object.assign`, its generally bad idea to use it in the constructors for anything else than scalar properties (or anything else than just a simple assignment). Also that final part in the test with `const result = (await orm.em.findOne(Member, member.id))!;` is a no-op, it won't reload anything, you query the entity by PK and it is already loaded, this won't touch the database at all.

Also, your changes in #3600 broke the change set computation, resulting in unwanted update queries even with the explicit collection API usage, weird the existing tests didn't catch this explicitly.

So in a nutshell, multiple bugs combined, one of them being introduced by your changes trying to fix another. What a nice day :D
